### PR TITLE
[Feature] Hdfs write fail, the data will write to localfile

### PR DESCRIPTION
**Pull request by jerqi**
### What changes were proposed in this pull request?
When we fail to write hdfs, we use localfile instead in multistorage.

### Why are the changes needed?
#56 we support mutlistorage read data, but hdfs is unstable sometimes. So we need a fallback mechanism that when data can't be written.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
New UT

### CODE REVIEW

File: server/src/main/java/com/tencent/rss/server/ShuffleFlushManager.java [hunk 1]
```diff
@@ -193,15 +192,29 @@ private void flushToFile(ShuffleDataFlushEvent event) {
...
           } catch (Exception e) {
             LOG.warn("Exception happened when write data for " + event + ", try again", e);
             ShuffleServerMetrics.counterWriteException.inc();
             Thread.sleep(1000);
           } finally {
             storage.unlockShuffleShared(RssUtils.generateShuffleKey(event.getAppId(), event.getShuffleId()));
           }
-          retry++;
-        }
+          event.increaseRetryTimes();
```
Comment (duanmeng) [inline server/src/main/java/com/tencent/rss/server/ShuffleFlushManager.java:203]: The fallback granularity is per event?

File: server/src/main/java/com/tencent/rss/server/ShuffleDataFlushEvent.java [hunk 2]
```diff
@@ -35,6 +36,7 @@
   private List<ShufflePartitionedBlock> shuffleBlocks;
   private Supplier<Boolean> valid = null;
   private ShuffleBuffer shuffleBuffer;
+  private AtomicInteger retryTimes = new AtomicInteger();
```
Comment (duanmeng) [inline server/src/main/java/com/tencent/rss/server/ShuffleDataFlushEvent.java:39]: final

File: server/src/main/java/com/tencent/rss/server/ShuffleFlushManager.java (hunk 1 above)
Comment (jerqi) [reply server/src/main/java/com/tencent/rss/server/ShuffleFlushManager.java:203]: Yes

File: server/src/main/java/com/tencent/rss/server/ShuffleDataFlushEvent.java (hunk 2 above)
Comment (jerqi) [reply server/src/main/java/com/tencent/rss/server/ShuffleDataFlushEvent.java:39]: OK

**Comment by colinmjj** [comment]
I think storage should has api `supportFallbak` as current implementation.
But how to fallback, it should be storage's action which should be wrapped in storage.
how about update interface `ShuffleWriteHandler.write(List)` -> 'ShuffleWriteHandler.write(Event)', the information like retry times can be got in write method.

**Comment by jerqi** [comment]
> I think storage should has api `supportFallbak` as current implementation. But how to fallback, it should be storage's action which should be wrapped in storage. how about update interface `ShuffleWriteHandler.write(List)` -> 'ShuffleWriteHandler.write(Event)', the information like retry times can be got in write method.

you means that we should have a multistorageWriteHanlder to fallback?

**Comment by jerqi** [comment]
> I think storage should has api `supportFallbak` as current implementation. But how to fallback, it should be storage's action which should be wrapped in storage. how about update interface `ShuffleWriteHandler.write(List)` -> 'ShuffleWriteHandler.write(Event)', the information like retry times can be got in write method.

Do you mean that Storage has api `supportFallback` instead StorageManager?

**Comment by jerqi** [comment]
All comments are addressed.

**Comment by jerqi** [comment]
GA failed because of flaky tests. And I rerun them.

File: coordinator/src/main/java/com/tencent/rss/coordinator/CoordinatorGrpcService.java [hunk 3]
```diff
@@ -135,7 +135,7 @@ public void heartbeat(
         .setRetMsg("")
         .setStatus(StatusCode.SUCCESS)
         .build();
-    LOG.debug("Got heartbeat from " + serverNode);
+    LOG.warn("Got heartbeat from " + serverNode);
```
Comment (colinmjj) [inline coordinator/src/main/java/com/tencent/rss/coordinator/CoordinatorGrpcService.java:138]: for 100+ shuffle server, there are too much heartbeat log

File: server/src/main/java/com/tencent/rss/server/ShuffleServerConf.java [hunk 4]
```diff
@@ -336,6 +336,13 @@
       .defaultValue(64L * 1024L * 1024L)
       .withDescription("For multistorage, the event size exceed this value, flush data  to cold storage");
 
+  public static final ConfigOption<Long> FALLBACK_TIMES = ConfigOptions
+      .key("rss.server.fallback.times")
```
Comment (colinmjj) [inline server/src/main/java/com/tencent/rss/server/ShuffleServerConf.java:340]: rss.server.multipleStorage.failedTimesBeforeFallback?

File: storage/src/main/java/com/tencent/rss/storage/common/Storage.java [hunk 5]
```diff
@@ -42,6 +42,8 @@
 
   ShuffleWriteHandler getOrCreateWriteHandler(CreateShuffleWriteHandlerRequest request) throws IOException;
 
+  CreateShuffleWriteHandlerRequest getCreateWriterHandlerRequest(String appId, int shuffleId, int partition);
```
Comment (colinmjj) [inline storage/src/main/java/com/tencent/rss/storage/common/Storage.java:45]: why this api is necessary?

Comment (jerqi) [reply storage/src/main/java/com/tencent/rss/storage/common/Storage.java:45]: When we write hdfs failed, we need use the request to create local handler.

File: server/src/main/java/com/tencent/rss/server/ShuffleServerConf.java (hunk 4 above)
Comment (jerqi) [reply server/src/main/java/com/tencent/rss/server/ShuffleServerConf.java:340]: OK

File: coordinator/src/main/java/com/tencent/rss/coordinator/CoordinatorGrpcService.java (hunk 3 above)
Comment (jerqi) [reply coordinator/src/main/java/com/tencent/rss/coordinator/CoordinatorGrpcService.java:138]: Ok

File: server/src/main/java/com/tencent/rss/server/ShuffleServerConf.java (hunk 4 above)
Comment (jerqi) [reply server/src/main/java/com/tencent/rss/server/ShuffleServerConf.java:340]: For configuration, we should use lower case letters.

File: storage/src/main/java/com/tencent/rss/storage/common/Storage.java (hunk 5 above)
Comment (colinmjj) [reply storage/src/main/java/com/tencent/rss/storage/common/Storage.java:45]: ok

File: server/src/main/java/com/tencent/rss/server/ShuffleServerConf.java (hunk 4 above)
Comment (colinmjj) [reply server/src/main/java/com/tencent/rss/server/ShuffleServerConf.java:340]: ok

File: server/src/main/java/com/tencent/rss/server/ShuffleServerConf.java [hunk 6]
```diff
@@ -336,6 +336,13 @@
       .defaultValue(64L * 1024L * 1024L)
       .withDescription("For multistorage, the event size exceed this value, flush data  to cold storage");
 
+  public static final ConfigOption<Long> FALLBACK_TIMES = ConfigOptions
```
Comment (colinmjj) [inline server/src/main/java/com/tencent/rss/server/ShuffleServerConf.java:339]: FALLBACK_TIMES -> FALLBACK_MAX_FAIL_TIMES

Comment (jerqi) [reply server/src/main/java/com/tencent/rss/server/ShuffleServerConf.java:339]: OK

**Comment by colinmjj** [comment]
LGTM

**Review by colinmjj — APPROVED** [review]

**Pull request merged**